exetera/core/field_importers.py

```diff
@@ -0,0 +1,344 @@
+}
+
+#========= ImporterDefinition, include Categorical, Numeric, , , Datetime =========
+
+class ImporterDefinition:
+    def __init__(self):
+        self._field_size = 0
+        self._importer = None
+
+
+class Categorical(ImporterDefinition):
+    def __init__(self, categories, value_type='int8', allow_freetext=False):
+        """
+        :param categories: dictionary that contain key/value pair for Categorical Field
+        :param value_type: value type in the dictionary. Default is 'int8'.
+        """
+        self._field_size = max([len(k) for k in categories.keys()])
+
+        self._importer = FIELD_MAPPING_TO_IMPORTER['categorical'](categories, value_type, allow_freetext)
+
+
+class Numeric(ImporterDefinition):
+    def __init__(self, dtype, invalid_value=0, validation_mode='allow_empty', create_flag_field='True', flag_field_name='_valid'):
+        if dtype in ('int', 'int8', 'int16', 'int32', 'int64', 'uint8', 'uint16', 'uint32', 'uint64'):
+            self._field_size = 20
+        elif dtype in ('float', 'float32', 'float64'):
+            self._field_size = 30
+        elif dtype == 'bool':
+            self._field_size = 5
+        else:
+            raise ValueError("Unrecognised numeric type '{}' in the field".format(dtype))
+
+        self._importer = FIELD_MAPPING_TO_IMPORTER['numeric'](dtype, invalid_value, validation_mode, create_flag_field, flag_field_name)
+
+
+class String(ImporterDefinition):
+    def __init__(self, fixed_length: int = None):
+        if fixed_length:
+            self._field_size = fixed_length
+        else:
+            self._field_size = 10 # guessing
+
+        self._importer = FIELD_MAPPING_TO_IMPORTER['string'](fixed_length)
+
+
+class DateTime(ImporterDefinition):
+    def __init__(self, create_day_field=False, create_flag_field=False):
+        self._field_size = 32
+        self._importer = FIELD_MAPPING_TO_IMPORTER['datetime'](create_day_field, create_flag_field)
+
+
+class Date(ImporterDefinition):
+    def __init__(self, create_flag_field=False):
+        self._field_size = 10
+        self._importer = FIELD_MAPPING_TO_IMPORTER['date'](create_flag_field)
+
+
+#============= Field Importers ============
+
+class CategoricalImporter:
+    def __init__(self, session, df:DataFrame, name:str, categories:Mapping[str, str], value_type:str='int8', timestamp=None):
+        if not isinstance(categories, dict):
+            raise ValueError("'categories' must be of type dict but is {} in the field '{}'".format(type(categories, name)))
+        elif len(categories) == 0:
+            raise ValueError("'categories' must not be empty in the field '{}'".format(name))
+
+        self.field = df.create_categorical(name, value_type, categories, timestamp, None)
+        self.byte_map = ops.get_byte_map(categories)
+        self.field_size = max([len(k) for k in categories])
+
+    def transform_and_write_part(self, column_inds, column_vals, column_offsets, col_idx, written_row_count):
```

Comment:
Can the importers themselves get a base class? Can 'transform_and_write_part' be renamed to something more succinct such as 'import_part'?

Reply:
what about 'convert_part'? just thinking 'import' is found in many places.

Reply:
I'd prefer import_part. It is more specific. convert_part could have other semantic meanings, such as converting already imported datatypes.

Reply:
Sure, will change it.